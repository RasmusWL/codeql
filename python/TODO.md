We have gotten quite a lot of QL classes now... and their names doesn't map very well to the call types they handle.

left some TODOs in the code right now.

First item on the agenda is to, in both (datamodel, calls/test.py, and the old callgraph testing), add empty subclass and see that calls do not get resolved... obviously the current handling of `C.foo(C, arg0)` is bad, since it's not class based, but function based... and that will not handle subclass shit...

Can definitly do a grouping that is (1) plain functions (2) method calls on classes (bound-method, classmethod, staticmethod, direct call of method (non-bound))
