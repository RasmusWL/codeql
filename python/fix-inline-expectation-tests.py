#!/usr/bin/env python3

import sys
from pathlib import Path
from dataclasses import MISSING, dataclass
from collections import defaultdict
from enum import Enum
from typing import List
import re

EXPECTATION_RE = r"#\s*\$.*$"

@dataclass
class FileSpec:
    filename: str
    start_line: int
    start_column: int
    end_line: int
    end_colum: int

    def __init__(self, filename, start_line, start_column, end_line, end_colum):
        self.filename = filename.strip()
        self.start_line = int(start_line)
        self.start_column = int(start_column)
        self.end_line = int(end_line)
        self.end_colum = int(end_colum)

class ProblemType(Enum):
    FIXED_MISSING = "Fixed missing result:"
    UNEXPECTED_RESULT = "Unexpected result: "
    MISSING_RESULT = "Missing result:"

class HandleMissing(Enum):
    REMOVE = "remove"
    ADD_TO_MISSING = "add"

@dataclass
class Problem:
    problem: ProblemType
    tag: str
    value: str

    def format_expectation(self) -> str:
        if self.value == "":
            return self.tag
        else:
            return f"{self.tag}={self.value}"

    @classmethod
    def from_input(cls, raw: str):
        raw = raw.strip()
        for pt in ProblemType:
            if raw.startswith(pt.value):
                break
        else:
            raise ValueError("Not a known problem type", raw)

        tag, value = raw[len(pt.value):].split("=")

        return cls(pt, tag, value)


def fix_concept_tests(ct_expected_path: Path, handle_missing: HandleMissing = HandleMissing.REMOVE):

    problems_in_file = defaultdict(lambda: defaultdict(list))

    lines = ct_expected_path.read_text().splitlines()

    if lines[0].startswith("|"):
        inline_expectation_lines = lines
    else:
        # there are multiple query predicates
        start = None
        end = None

        for i, line in enumerate(lines):
            if start is None:
                if line == "failures":
                    start = i + 1
            else:
                if not line.startswith("|"):
                    end = i
                    break

        assert start is not None
        inline_expectation_lines = lines[start:end]

    for line in inline_expectation_lines:
        _, file_spec_raw, _, problem_raw, _ = line.split("|")

        spec = FileSpec(*file_spec_raw.split(":"))

        if spec.start_line != spec.end_line:
            print(f"skipping line {spec.start_line}-{spec.end_line} of {spec.filename} since this script doesn't have multiline support: {line}", file=sys.stderr)
            continue

        problem = Problem.from_input(problem_raw)

        problems_in_file[spec.filename][spec.start_line].append(problem)

    for filename, problems_on_line in problems_in_file.items():
        with ct_expected_path.with_name(filename).open('r+') as f:
            lines = f.readlines()

            for line_num, problems in problems_on_line.items():
                line_index = line_num-1
                problems: List[Problem]
                stuff_we_can_handle = [ProblemType.UNEXPECTED_RESULT]
                if not all(p.problem in stuff_we_can_handle for p in problems):
                    print(f"Skipping line {line_num} of {filename} since not all problems are {stuff_we_can_handle}", file=sys.stderr)
                    continue

                line = lines[line_index]
                if (match := re.search(EXPECTATION_RE, line)):
                    new_expectations = [match.group(0)]
                    line = line[:match.start()]
                else:
                    new_expectations = [" # $"]
                    line = line.rstrip()

                for problem in problems:
                    if problem.problem == ProblemType.UNEXPECTED_RESULT:
                        to_add = problem.format_expectation()
                        if to_add not in new_expectations:
                            new_expectations.append(to_add)
                    else:
                        raise Exception("problem type not handled", problem)

                new_expecatations_formatted = " ".join(new_expectations)
                lines[line_index] = line + new_expecatations_formatted

            f.truncate()
            f.seek(0)
            for line in lines:
                f.write(line.rstrip("\n"))
                f.write("\n")


if __name__ == "__main__":
    ct_in = Path(sys.argv[1])
    fix_concept_tests(ct_in)
