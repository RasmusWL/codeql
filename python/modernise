#!/bin/bash

set -e

CURRENTDIR=$(pwd)
SCRIPTDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
PATTERNS="$SCRIPTDIR/modernise-patterns"

name="$1"
test=$(readlink -f "$2")
files="${@:3}"
auto_branch_name="python-modernise-$(echo $name | tr ' ' '-')"

if [ -z $SEMMLE_CODE ]; then
    echo "\$SEMMLE_CODE must be set to your 'code' repo"
    exit 1
fi

# TODO: Understand the case syntax :D

function should_run {
    echo -n "$@? "
    read -p "[Yn]" yn
    case $yn in
        [Nn]* ) /bin/false;;
        * ) /bin/true;;
    esac
}

function should_run_false {
    echo -n "$@? "
    read -p "[yN]" yn
    case $yn in
        [Yy]* ) /bin/true;;
        * ) /bin/false;;
    esac
}

for file in $files; do
    echo $file
done


if should_run "new branch (${auto_branch_name})"; then
    git checkout -b ${auto_branch_name}
fi

if should_run "auto format"; then
    pyqlformat $files

    if should_run "commit auto format"; then
    git commit -am "Python: Autoformat ${name}"
    fi
fi


if should_run "modernise (regexp replace)"; then
    repren -p "$PATTERNS" $files
    repren -p "$PATTERNS" $files
    repren -p "$PATTERNS" $files

    for file in $files; do
        rm -f $file.orig
    done
fi

if should_run "run tests on ${test}"; then
    if [[ -d "${test}" ]]; then
        test_folder="${test}"
    else
        test_folder=$(dirname "${test}")
    fi

    while :; do
        cd $SEMMLE_CODE && ./target/jars/qltest --threads 8 "${test}" || ~/bin/qlicdiff "${test_folder}"
        if should_run_false "accept results"; then
            rename -f -e 's/\.actual$/.expected/' ${test_folder}/*.actual
            break
        fi
    done
fi

cd "${CURRENTDIR}"

if should_run "commit modernise"; then
    git commit -am "Python: Modernise ${name}"
fi

if should_run "make pull-request"; then
    current_branch_name=$(git rev-parse --abbrev-ref HEAD)
    git push origin ${current_branch_name}
    xdg-open "https://github.com/RasmusWL/ql/pull/new/${current_branch_name}"
    git checkout master
fi
