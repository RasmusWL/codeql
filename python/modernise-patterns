\(FunctionObject\)\.getFunction\(\)	(FunctionValue).getScope()
\.getPyClass\(\)	.getScope()

# -> Value::named
ModuleObject::named\("(.*)"\).attr\("(.*)"\)	Value::named("\1.\2")
ModuleObject::named\("(.*)"\)	Value::named("\1")
Module::named\("(.*)"\).attr\("(.*)"\)	Value::named("\1.\2")

# Object -> Value
\bModuleObject\b	ModuleValue
\bFunctionObject\b	FunctionValue
\bClassObject\b	ClassValue
\bStringObject\b	StringValue
\bNumericObject\b	NumericValue
\bTupleObject\b	TupleValue
\bObject\b	Value

\bPyFunctionObject\b	PythonFunctionValue

\.lookupAttribute\(	.lookup(

# refersTo -> pointsTo
\.refersTo\(_, (.*), _\)	.pointsTo().getClass() = \1
\.refersTo\((.*)\)	.pointsTo(\1)

\.getASuperType\(\)	.getABaseType+()
