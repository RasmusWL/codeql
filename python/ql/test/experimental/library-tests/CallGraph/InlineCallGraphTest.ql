import python
import TestUtilities.InlineExpectationsTest
private import semmle.python.dataflow.new.internal.DataFlowDispatch as TT

/** A call graph edge resolved based on Type Trackers */
predicate pointsToCallEdge(CallNode call, Function callable) {
  exists(PythonFunctionValue funcValue |
    funcValue.getScope() = callable and
    call = funcValue.getACall()
  )
}

/** A call graph edge resolved based on Type Trackers */
predicate typeTrackerCallEdge(CallNode call, Function callable) {
  TT::TFunction(callable) = TT::viableCallable(TT::TNormalCall(call))
}

/** Holds if the call edge is from a class call. */
predicate typeTrackerClassCall(CallNode call, Function callable) {
  exists(TT::ClassCall cc |
    cc.getNode() = call and
    TT::TFunction(callable) = cc.getCallable()
  )
}

class CallGraphTest extends InlineExpectationsTest {
  CallGraphTest() { this = "CallGraphTest" }

  override string getARelevantTag() { result in ["pt", "tt"] }

  override predicate hasActualResult(Location location, string element, string tag, string value) {
    exists(location.getFile().getRelativePath()) and
    exists(CallNode call, Function target |
      tag = "tt" and
      typeTrackerCallEdge(call, target)
      or
      tag = "pt" and
      pointsToCallEdge(call, target)
    |
      location = call.getLocation() and
      element = call.toString() and
      if call.getLocation().getFile() = target.getLocation().getFile()
      then value = betterQualName(target)
      else value = target.getLocation().getFile() + ":" + betterQualName(target)
    )
  }
}

bindingset[func]
string betterQualName(Function func) {
  // note: `target.getQualifiedName` for Lambdas is just "lambda", so is not very useful :|
  not func.isLambda() and
  result = func.getQualifiedName()
  or
  func.isLambda() and
  result =
    "lambda[" + func.getLocation().getFile().getShortName() + ":" +
      func.getLocation().getStartLine() + ":" + func.getLocation().getStartColumn() + "]"
}

query predicate debug_callableNotUnique(Function callable, string message) {
  exists(Function f |
    f != callable and
    f.getQualifiedName() = callable.getQualifiedName() and
    f.getLocation().getFile() = callable.getLocation().getFile()
  ) and
  message =
    "Qualified function name '" + callable.getQualifiedName() +
      "' is not unique within its file. Please fix."
}

query predicate pointsTo_found_typeTracker_notFound(CallNode call, string qualname) {
  exists(Function target |
    pointsToCallEdge(call, target) and
    not typeTrackerCallEdge(call, target) and
    qualname = betterQualName(target)
  )
}

query predicate pointsTo_notFound_typeTracker_found(CallNode call, string qualname) {
  exists(Function target |
    not pointsToCallEdge(call, target) and
    typeTrackerCallEdge(call, target) and
    qualname = betterQualName(target) and
    // We filter out result differences for points-to and type-tracking for class calls,
    // since otherwise it gives too much noise (these are just handled differently
    // between the two).
    not typeTrackerClassCall(call, target)
  )
}
