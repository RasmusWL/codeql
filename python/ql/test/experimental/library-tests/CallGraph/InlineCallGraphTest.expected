failures
debug_callableNotUnique
| code/class_properties.py:7:5:7:18 | Function arg | Qualified function name 'Prop.arg' is not unique within its file. Please fix. |
| code/class_properties.py:12:5:12:25 | Function arg | Qualified function name 'Prop.arg' is not unique within its file. Please fix. |
| code/class_properties.py:17:5:17:18 | Function arg | Qualified function name 'Prop.arg' is not unique within its file. Please fix. |
pointsTo_found_typeTracker_notFound
| code/class_attr_assign.py:10:9:10:27 | ControlFlowNode for Attribute() | my_func |
| code/class_attr_assign.py:11:9:11:25 | ControlFlowNode for Attribute() | my_func |
| code/class_attr_assign.py:26:9:26:25 | ControlFlowNode for Attribute() | DummyObject.method |
| code/class_super.py:50:1:50:6 | ControlFlowNode for Attribute() | outside_def |
| code/conditional_in_argument.py:18:5:18:11 | ControlFlowNode for Attribute() | X.bar |
| code/func_defined_outside_class.py:21:1:21:11 | ControlFlowNode for Attribute() | A.foo |
| code/func_defined_outside_class.py:22:1:22:15 | ControlFlowNode for Attribute() | outside |
| code/func_defined_outside_class.py:24:1:24:14 | ControlFlowNode for Attribute() | outside_sm |
| code/func_defined_outside_class.py:25:1:25:14 | ControlFlowNode for Attribute() | outside_cm |
| code/func_defined_outside_class.py:38:11:38:21 | ControlFlowNode for _gen() | B._gen |
| code/func_defined_outside_class.py:39:11:39:21 | ControlFlowNode for _gen() | B._gen |
| code/func_defined_outside_class.py:42:1:42:7 | ControlFlowNode for Attribute() | B._gen.func |
| code/func_defined_outside_class.py:43:1:43:7 | ControlFlowNode for Attribute() | B._gen.func |
| code/type_tracking_limitation.py:8:1:8:3 | ControlFlowNode for x() | my_func |
typeTracker_found_pointsTo_notFound
| code/callable_as_argument.py:29:5:29:12 | ControlFlowNode for Attribute() | test_class.InsideTestFunc.sm |
| code/class_more_mro2.py:18:9:18:21 | ControlFlowNode for Attribute() | A.foo |
| code/class_more_mro2.py:21:1:21:8 | ControlFlowNode for Attribute() | A.foo |
| code/class_more_mro.py:24:9:24:21 | ControlFlowNode for Attribute() | A.foo |
| code/class_more_mro.py:34:1:34:16 | ControlFlowNode for Attribute() | A.foo |
| code/class_super.py:43:9:43:21 | ControlFlowNode for Attribute() | A.bar |
| code/class_super.py:44:9:44:27 | ControlFlowNode for Attribute() | A.bar |
| code/class_super.py:63:1:63:18 | ControlFlowNode for Attribute() | A.foo |
| code/class_super.py:78:9:78:28 | ControlFlowNode for Attribute() | A.foo |
| code/class_super.py:81:1:81:12 | ControlFlowNode for Attribute() | C.foo_on_A |
| code/class_super.py:92:9:92:21 | ControlFlowNode for Attribute() | X.foo |
| code/class_super.py:97:9:97:21 | ControlFlowNode for Attribute() | X.foo |
| code/class_super.py:97:9:97:21 | ControlFlowNode for Attribute() | Y.foo |
| code/class_super.py:101:1:101:7 | ControlFlowNode for Attribute() | Z.foo |
| code/class_super.py:108:1:108:8 | ControlFlowNode for Attribute() | Z.foo |
| code/def_in_function.py:22:5:22:11 | ControlFlowNode for Attribute() | test.A.foo |
| code/nested_class.py:83:9:83:16 | ControlFlowNode for Attribute() | X.class_def_in_func.Y.meth |
| code/underscore_prefix_func_name.py:14:5:14:19 | ControlFlowNode for some_function() | some_function |
